theano/gpuarray/opt.py

```diff
@@ -210,6 +210,52 @@ def apply(self, fgraph):
                     0, 'fast_run', 'fast_compile', 'merge')
 
 
+class GraphToGPU(Optimizer):
+    """
+    Transfer the graph as a whole to GPU instead of replacing nodes
+    """
+    def add_requirements(self, fgraph):
+        fgraph.attach_feature(toolbox.ReplaceValidate())
+
+    def apply(self, fgraph):
+        mapping = {}
+        move_to_GPU = True
+
+        # Building a new graph
+        for i in fgraph.inputs:
+            mapping[i] = GpuFromHost(None)(i)
```

Comment:
It depend how you inspect the node.

A node have 2 lists of variables. One for its inputs and one for its outputs (most of the time 1, but there is a few op that have more then 1 output)

If you check all nodes in the graph and check only the inputs variables of the nodes, then yes, you will see all inputs of the graph (except if there are some that aren't used). But you won't see the output variable of the graph.
If you only check the outputs of each nodes, then you will see the outputs variable of the graph, but not the input of the graph.

Here, we want to see all variables. If we start by the input of the graph + the outputs of each nodes, this make sure we see all variables of the graph.

The code in the following for on each node, only add to the mapping the outputs of nodes, not the inputs of each nodes (they should already be there)